Association_Mining/blog.md

```diff
@@ -0,0 +1,219 @@
+
+```
+data['Invoice'] = data['Invoice'].astype('str') 
+data = data[~data['Invoice'].str.contains('C')] 
+```
+
+In this code snippet, we convert the *Invoice* column into string type before finding if it contains a 'C'.
+
+##### Applying Apriori Algorithm
+
+The next step is to apply the Apriori algorithm on the dataset. To do so, we can use the apriori function that we imported from the *mlxtend* library.
+
+The sales and thus, the association rules will be different in different countries. For example, people in Saudi Arabia will probably buy less number of winter clothes than the people in Canada. So we will make baskets (subset of items) according to countries.
+Let's start with finding what are the countries which are included in the data.
+
+```
+data.Country.unique()
+```
+
+We can create baskets for each country by executing the following code:
+
+```
+basket_USA = (data[data['Country'] =="USA"] 
+          .groupby(['Invoice', 'Description'])['Quantity'] 
+          .sum().unstack().reset_index().fillna(0) 
+          .set_index('Invoice'))
+```
+
+Here, we have filtered the data according to the country. We need to group the data according to *Invoice* number and *Description* and structure the data such that for each invoice, we can give a value *(Quantity)* to each item based on whether it is bought. If a certain item is bought in a transaction, we put the *quantity* or else a *0*. 
+
+Because, we are not concerned with the Quantity of the product sold, we convert the values to *0* or *1* indicating *False* or *True* respectively. It shows if the product is sold in a particular transaction.
+
+```
+def hot_encode(x): 
+    if(x<= 0): 
+        return 0
+    if(x>= 1): 
+        return 1
+        
+basket_encoded = basket_USA.applymap(hot_encode) 
+basket_USA = basket_encoded
+```
+
+Using the apriori function, and also assuming the support threshold to be *0.06*, we find the frequent item sets.
+
+```
+frq_items = apriori(basket_USA, min_support = 0.06, use_colnames = True)
+```
+
+To find the association rules, we use the *association_rules()* function. We pass the frequent item sets as one of the parameters. We use the **Lift** metric and set the minimum threshold to be 1.2.
+
+```
+rules = association_rules(frq_items, metric ="lift", min_threshold = 1.2)
+rules = rules.sort_values(['confidence', 'lift'], ascending =[False, False]) 
+```
+
+We also sort the rules based on the confidence value and the lift metric value such that the rules with higher confidence and lift are shown on the top.
+
+Let us see how the rules table looks. Execute the following code and you should get an output similar to the image shown below.
+
+```
+rules.head(6)
+```
+
+![rules_head output](/Association_Mining/images/rules_head.png)
+
+For example, here it shows that if a customer in USA buys a *Mini Paint Set Vintage*, he/she will most likely buy *60 Cake Cases Dolly Girl Design*.

```

Comment:
add a sentence before this one saying what the table shows overall, something like: "This table shows the relationships between products" (maybe use more detail, you now how this works better than I do)